### PR TITLE
Add Dawson's Integral.

### PR DESCRIPTION
**Pull request by ethanbarry**
This commit adds the Dawson function, using the implementation in Numerical Recipes, a helper function called `sign()`, and a test file under the bin directory.

### CODE REVIEW

**Comment by ethanbarry** [comment]
Hey! I saw this crate while digging through peroxide's documentation, and thought I'd contribute this little port of the Numerical Recipes source. I have a couple things to note, though.
1. The code includes an unused boolean flag and `if` statement. We could remove this and always initialize the array `c`.
2. The approximation is accurate to around 2E-7, as the book says. In the test case I provided, it's a little better. But in the case of larger arguments, like F(12), the approximation doesn't even come within 4 decimal places of the correct answer. We could explore a more accurate approximation...

Please let me know what I can do to make this commit better! Thanks! :)

File: src/lib.rs [hunk 1]
```diff
@@ -500,6 +548,63 @@ pub fn invbetai(p: f64, a: f64, b: f64) -> f64 {
...
+    let mut d2: f64;
+    let mut e1: f64;
+    let e2: f64;
+    let mut sum: f64;
+    let x2: f64;
+    let xp: f64;
+    let xx: f64;
+    let ans: f64;
+
+    if init {
```
Comment (Axect) [inline src/lib.rs:575]: The `init` variable and its associated `if` statement seem to be implemented incorrectly:

- `init` is always `true` when the `if` statement is reached
- The `if` block will always execute
- Setting `init = false` inside the block has no effect as it's never used again

Please review this part of the implementation and update it accordingly. If lazy initialization was intended, consider using a static variable or a different approach to achieve that goal.

**Review by Axect — CHANGES_REQUESTED** [review]
Thank you for your contribution! Here are my thoughts on your pull request:

- I'm impressed and appreciative of your pure Rust implementation of the Dawson integral. It's a valuable addition to the project.

- I've verified that the tests are working correctly, which is great.

- As mentioned in my inline comment, I noticed the use of an unnecessary mutable `init` variable and an `if` statement within the `dawson` function that don't seem to serve a purpose.

	- If the intention is simply initialization, I recommend removing the `if` statement and directly initializing the values at the start of the function.
	
- Once this small issue is addressed, I'll be happy to merge this pull request.

Thank you again for your contribution. Looking forward to the updated version!

**Comment by Axect** [comment]
> EDIT: Aaaand I just realized rust-analyzer was running rustfmt on the source code, and reformatted a ton of lines. Let me see if I can fix that...

No worries about the rustfmt changes! Actually, having the code formatted consistently is beneficial for the project. We appreciate that rust-analyzer automatically applied these formatting rules. It helps maintain a uniform code style across the project, which improves readability and maintainability. So, there's no need to "fix" the formatting changes - they're welcome improvements!

**Comment by ethanbarry** [comment]
Wonderful! Thank you so much for reviewing it!

Regarding the `init` flag, this is what appears in the C++ source from Numerical Recipes—a boolean that is always true when the if statement is entered. I also see no purpose to having it in the source; my copy of the book may have simply had a mistake. I'll remove it when I can access my computer!

-------- Original Message --------
On Aug 1, 2024, 7:02 PM, Tae-Geun Kim < ***@***.***> wrote:
@Axect requested changes on this pull request.
Thank you for your contribution! Here are my thoughts on your pull request:
I'm impressed and appreciative of your pure Rust implementation of the Dawson integral. It's a valuable addition to the project.
I've verified that the tests are working correctly, which is great.
As mentioned in my inline comment, I noticed the use of an unnecessary mutable init variable and an if statement within the dawson function that don't seem to serve a purpose.
If the intention is simply initialization, I recommend removing the if statement and directly initializing the values at the start of the function.
Once this small issue is addressed, I'll be happy to merge this pull request.
Thank you again for your contribution. Looking forward to the updated version!
In src/lib.rs:
> + let mut d2: f64; + let mut e1: f64; + let e2: f64; + let mut sum: f64; + let x2: f64; + let xp: f64; + let xx: f64; + let ans: f64; + + if init { + init = false; + for i in 0..NMAX { + c[i] = f64::exp(-((2. * i as f64 + 1.) * h).powf(2.)); + } + } +
The init variable and its associated if statement seem to be implemented incorrectly:
init is always true when the if statement is reached
The if block will always execute
Setting init = false inside the block has no effect as it's never used again
Please review this part of the implementation and update it accordingly. If lazy initialization was intended, consider using a static variable or a different approach to achieve that goal.
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.￼Message ID: ***@***.***>
[ { ***@***.***": "http://schema.org", ***@***.***": "EmailMessage", "potentialAction": { ***@***.***": "ViewAction", "target": "[https://github.com](<a href=)/Axect/puruspe/pull/9#pullrequestreview-2214261355">https://github.com/Axect/puruspe/pull/9#pullrequestreview-2214261355", "url": "[https://github.com](<a href=)/Axect/puruspe/pull/9#pullrequestreview-2214261355">https://github.com/Axect/puruspe/pull/9#pullrequestreview-2214261355", "name": "View Pull Request" }, "description": "View this Pull Request on GitHub", "publisher": { ***@***.***": "Organization", "name": "GitHub", "url": "https://github.com" } } ]

File: src/lib.rs [hunk 2]
```diff
@@ -500,6 +548,63 @@ pub fn invbetai(p: f64, a: f64, b: f64) -> f64 {
...
+        init = false;
+        for i in 0..NMAX {
+            c[i] = f64::exp(-((2. * i as f64 + 1.) * h).powf(2.));
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (ethanbarry) [reply src/lib.rs:577]: Okay, I think this takes care of it! I used `git commit --amend` on this one, in case that changes anything on your end...

**Review by Axect — APPROVED** [review]

**Comment by Axect** [comment]
Thank you for addressing the feedback and updating the pull request. The changes look great:

- Removed unnecessary `init` variable and `if` statement
- Code is now more efficient and straightforward

Your Dawson integral implementation is a valuable addition. The code looks good and tests are passing.

I'm approving this PR and will merge it shortly. Thanks for your excellent work!

**Comment by ethanbarry** [comment]
No, thank you! This was fun!

**Pull request merged**